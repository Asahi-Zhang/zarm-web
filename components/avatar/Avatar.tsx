import React, { Component, HTMLAttributes, CSSProperties } from 'react';
import classnames from 'classnames';
import { PropsType, StateType } from './PropsType';

class Avatar extends Component<PropsType & HTMLAttributes<HTMLSpanElement>, StateType> {
  static defaultProps: PropsType = {
    prefixCls: 'zw-avatar',
    shape: 'circle',
    src: '',
    size: 'md',
    alt: '',
  };

  private avatarChildrenNode: HTMLElement;

  private avatarWrapperNode: HTMLElement;

  state: StateType = {
    childrenScale: 1,
  };


  componentDidMount() {
    this.setChildrenScale();
  }

  componentDidUpdate(prevProps: PropsType) {
    if (this.props !== prevProps) {
      this.setChildrenScale();
    }
  }

  setChildrenScale = () => {
    if (!this.avatarChildrenNode || !this.avatarWrapperNode) {
      return;
    }
    const avatarChildrenWidth = this.avatarChildrenNode.offsetWidth;
    const avatarWrapperWidth = this.avatarWrapperNode.offsetWidth;
    if (!avatarChildrenWidth || !avatarWrapperWidth) {
      return;
    }
    const childrenScale = (avatarWrapperWidth - 8) / avatarChildrenWidth;
    this.setState({
      childrenScale: (childrenScale < 1)
        ? childrenScale
        : 1,
    });
  };

  render() {
    const {
      prefixCls,
      style,
      size,
      shape,
      src,
      alt,
      children,
      className,
      ...others
    } = this.props;
    const { childrenScale } = this.state;

    const hasFontSizeStyle = (style && style.fontSize);
    const hasImage = (src && src.trim() !== '');
    const hasString = (typeof children === 'string');

    const cls = classnames({
      [prefixCls!]: true,
      [className!]: className,
      [`${prefixCls}--${size}`]: typeof size === 'string' && size,
      [`${prefixCls}--${shape}`]: true,
    });

    const clsImage = classnames({ [`${prefixCls}--image`]: hasImage });
    const clsString = classnames({ [`${prefixCls}--string`]: hasString || hasImage });
    const childrenTransformStr = `scale(${childrenScale}) translateX(-50%)`;
    const spanStyle: CSSProperties = hasFontSizeStyle
      ? {} : {
        transform: childrenTransformStr,
        WebkitTransform: childrenTransformStr,
        msTransform: childrenTransformStr,
      };

    return (
      <span
        style={style}
        className={cls}
        {...others}
        ref={(node: HTMLElement) => {
          this.avatarWrapperNode = node;
        }}
      >
        {hasImage && <img src={src} alt={alt} className={clsImage} />}
        {(!hasImage) && hasString && (
          <span
            className={clsString}
            style={spanStyle}
            ref={(node: HTMLElement) => {
              this.avatarChildrenNode = node;
            }}
          >
            {children}
          </span>
        )}
      </span>
    );
  }
}

export default Avatar;
