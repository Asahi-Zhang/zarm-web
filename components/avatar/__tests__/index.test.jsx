import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Avatar from '../index';

describe('Avatar', () => {
  it('renders Avatars of different size correctly', () => {
    const wrapper = render(
      <div>
        <Avatar size={80} icon="time" />
        <Avatar size="xl" icon="time" />
        <Avatar size="lg" icon="time" />
        <Avatar icon="time" />
        <Avatar size="sm" icon="time" />
        <Avatar size="xs" icon="time" />
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars of different shape correctly', () => {
    const wrapper = shallow(
      <div>
        <Avatar size="circle" />
        <Avatar size="square" />
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars of different type correctly', () => {
    const wrapper = render(
      <div>
        <Avatar icon="time" />
        <Avatar>U</Avatar>
        <Avatar>USER</Avatar>
        <Avatar src="http://img95.699pic.com/element/40044/5588.png_860.png" alt="my avatar" onError={() => console.log('load error')} />
        <Avatar style={{ color: '#2db7f5', backgroundColor: 'rgb(249, 232, 8)' }}>U</Avatar>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars correctly when set src, icon and string', () => {
    const wrapper = render(
      <div>
        <Avatar src="http://img95.699pic.com/element/40044/5588.png_860.png" icon="time">USER</Avatar>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
