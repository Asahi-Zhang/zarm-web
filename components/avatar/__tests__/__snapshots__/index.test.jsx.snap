// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders Avatars correctly when set src, icon and string 1`] = `
<div>
  <span
    class="za-avatar is-circle has-icon has-image has-string"
  >
    <img
      alt=""
      src="http://img95.699pic.com/element/40044/5588.png_860.png"
    />
  </span>
</div>
`;

exports[`Avatar renders Avatars of different shape correctly 1`] = `
<div>
  <Avatar
    alt=""
    icon=""
    prefixCls="za-avatar"
    shape="circle"
    size="circle"
    src=""
  />
  <Avatar
    alt=""
    icon=""
    prefixCls="za-avatar"
    shape="circle"
    size="square"
    src=""
  />
</div>
`;

exports[`Avatar renders Avatars of different size correctly 1`] = `
<div>
  <span
    class="za-avatar is-circle has-icon"
    style="width:80px;height:80px;line-height:80px;font-size:40px"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--xl is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--lg is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--sm is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--xs is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
</div>
`;

exports[`Avatar renders Avatars of different type correctly 1`] = `
<div>
  <span
    class="za-avatar is-circle has-icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar is-circle has-string"
  >
    <span
      style="position:absolute;transform:scale(1);left:calc(50% - 4.5px )"
    >
      U
    </span>
  </span>
  <span
    class="za-avatar is-circle has-string"
  >
    <span
      style="position:absolute;transform:scale(0.7);left:calc(50% - 18px )"
    >
      USER
    </span>
  </span>
  <span
    class="za-avatar is-circle has-image"
  >
    <img
      alt="my avatar"
      src="http://img95.699pic.com/element/40044/5588.png_860.png"
    />
  </span>
  <span
    class="za-avatar is-circle has-string"
    style="color:#2db7f5;background-color:rgb(249, 232, 8)"
  >
    <span
      style="position:absolute;transform:scale(1);left:calc(50% - 4.5px )"
    >
      U
    </span>
  </span>
</div>
`;
